### ISSUE TITLE
Hybrid filtering - report any issues (or improvements) here

### ISSUE THREAD

**Issue by oddMLan** [gonetz/GLideN64#2116]
This filter attempts to solve a problem I've noticed in some games, where a texture is not filtered (nearest neighbor), yet since it's not resized by an integer factor, it causes some glitches. Bilinear filtering doesn't present this problem, but it makes the texture blurry, when it was intended to look crisp. Pictures are better than words:

Before and after
![before](https://user-images.githubusercontent.com/7278372/67791800-53f0a000-fa35-11e9-86a7-2ad26bfb7bef.png)
![after](https://user-images.githubusercontent.com/7278372/67791803-5652fa00-fa35-11e9-921e-ee0264c4ea2f.png)

Running in 800x600 resolution. Please ignore the minor tiling issue at the bottom, it's unrelated. But if you focus on the icons in the HUD, you can notice that before, the star didn't look quite right. In fact, it looks angry!

It also solves many minor issues such as this (when running in fullscreen):
Before and after
![GLideN64_PAPER_MARIO_007](https://user-images.githubusercontent.com/7278372/67792156-ef821080-fa35-11e9-9126-b8b5bd95a423.png)
![GLideN64_PAPER_MARIO_005](https://user-images.githubusercontent.com/7278372/67792202-fe68c300-fa35-11e9-83c6-3600f137ab2c.png)

You see some small yet noticeable artifacts in the Nintendo logo. Can be a bit distracting when playing in fullscreen.

Some games run at non-standard resolutions natively, and will never look right unless native or a factor of it is used.

Before
![GLideN64_RUGRATSSCAVENGERHUNT_001](https://user-images.githubusercontent.com/7278372/67792389-4d165d00-fa36-11e9-821f-e2105a273785.png)
![GLideN64_RUGRATSSCAVENGERHUNT_003](https://user-images.githubusercontent.com/7278372/67792404-51db1100-fa36-11e9-8111-5752664c1ca9.png)
After
![GLideN64_RUGRATSSCAVENGERHUNT_007](https://user-images.githubusercontent.com/7278372/67792425-5acbe280-fa36-11e9-8749-52375a660983.png)
![GLideN64_RUGRATSSCAVENGERHUNT_008](https://user-images.githubusercontent.com/7278372/67792438-615a5a00-fa36-11e9-92fc-13ca146dd920.png)

It almost looks bilinear filtered, but that's because none of the texels match the output resolution well. But it's better than the jagged mess that it was before.

So, the technical part about how this filter works, comes from this article. 
https://colececil.io/blog/2017/scaling-pixel-art-without-destroying-it/
Basically it tries to apply interpolation only to pixels that don't align with the pixel grid, preserving much of the original detail as possible. In fact, let's say a pixel represents five units on the screen, it'll only apply smoothing to 1 of those units.

Huge thanks to DKO for translating the original idea to clean GLSL code. Many thanks to @gonetz for assisting us with the GLideN64 integration.

Lastly please note, this doesn't solve #2082 since the blurry textures are marked as bilinear filtered. This filter in theory could be applied to all texture fetches in a #483 fashion, if one desired. I think it would be great for #2088 too! It will look good at any resolution without having uneven pixel row artifacts.

**Comment by oddMLan** [comment]
when testing this make sure you have native resolution or multiples of native disabled: you must use "match output resolution". native resolution never had this issue where the texels don't match

**Comment by DonelBueno** [comment]
@oddMLan Can you provide a test build without commit 0338f27 (the one from #2097)? Thanks in advance.

**Comment by oddMLan** [comment]
Test build without commit 0338f27 as requested by @DonelBueno 
I confirm that this also gets rid of bug #2112 
[GLideN64.zip](https://github.com/gonetz/GLideN64/files/3800935/GLideN64.zip)

**Comment by oddMLan** [comment]
Other improvements at default resolution (640x480)

before
![GLideN64_MarioTennis_003](https://user-images.githubusercontent.com/7278372/68077056-6bec5a80-fd7a-11e9-9478-91d9a90e2433.png)
![GLideN64_MarioTennis_004](https://user-images.githubusercontent.com/7278372/68077055-63941f80-fd7a-11e9-98d2-61cdddfd7b62.png)
after
![GLideN64_MarioTennis_000](https://user-images.githubusercontent.com/7278372/68077047-4f502280-fd7a-11e9-9ac9-46820c6544a8.png)
![GLideN64_MarioTennis_001](https://user-images.githubusercontent.com/7278372/68077060-73abff00-fd7a-11e9-8810-1d495c71917b.png)

**Comment by oddMLan** [comment]
improvement at 800x600 resolution

before
![GLideN64_THE_LEGEND_OF_ZELDA_009](https://user-images.githubusercontent.com/7278372/68077138-d225ad00-fd7b-11e9-8c6c-d75a2c24f7fe.png)

after
![GLideN64_THE_LEGEND_OF_ZELDA_010](https://user-images.githubusercontent.com/7278372/68077139-dc47ab80-fd7b-11e9-8707-740983e454ac.png)

**Comment by oddMLan** [comment]
if I disable checking for opaque pixels, the zelda logo looks great:

![GLideN64_THE_LEGEND_OF_ZELDA_012](https://user-images.githubusercontent.com/7278372/68077232-680e0780-fd7d-11e9-88bb-d637eaabf813.png)

but black artifacts are added to text in SM64:
![GLideN64_SUPER_MARIO_64_011](https://user-images.githubusercontent.com/7278372/68077253-cf2bbc00-fd7d-11e9-8840-0d9c2ba4a8b6.png)

@gonetz what's different about the texture parameters in these two cases? is it the cycle mode? or the bit depth? ideally I want to disable filtering only when knowing that the texture only supports 1 bit alpha. so that the zelda logo is improved without affecting textures like the SM64 text.

**Comment by oddMLan** [comment]
even then, there are some cases simply ignoring ignoring texels that aren't fully opaque don't get you a decent result:

in Mystical Ninja Starring Goemon, at 800x600 resolution, text shape looks wrong:
![GLideN64_MYSTICAL_NINJA_001](https://user-images.githubusercontent.com/7278372/68077460-9d1c5900-fd81-11e9-8b98-dd7fb4971d7b.png)

but if you filter all texels, text shape looks correct, except for those black artifacts due to the filtering:
![GLideN64_MYSTICAL_NINJA_000](https://user-images.githubusercontent.com/7278372/68077468-b9b89100-fd81-11e9-8caa-3c56128aa7e4.png)

What would happen if we treated all textures as 32-bit textures like the Zelda logo? That way the "black artifacts" would blend with the background, instead of black. Unless it is a CYCLE_MODE matter. Or something more complicated than that.

What's more significant, is that in the same game. When the logo is zooming out, and it sits for half a second, the logo doesn't have a black outline. After what I assume changes the texture parameters, the logo gets a black outline. .webm link: https://drive.google.com/uc?id=1KMwhjN4FQHhH9Fm3tRtHOHvRwbljLhbK

Is there a way that the plugin treats all textures as if they had full alpha so they always blend correctly with the background at non-native resolutions? I assume this is not always desired, for example the lens of truth effect. Although cosmetically it would look better, it would break the depth detection.

**Comment by oddMLan** [comment]
if I implement color bleeding to the filter, text in Mystical Ninja looks much better at 800x600

![GLideN64_MYSTICAL_NINJA_003](https://user-images.githubusercontent.com/7278372/68079609-740fbe80-fda9-11e9-82dc-506cfab82caf.png)

but zelda oot logo looks much worse
![GLideN64_THE_LEGEND_OF_ZELDA_013](https://user-images.githubusercontent.com/7278372/68079622-973a6e00-fda9-11e9-89e8-09455353fa90.png)

however putting if (uEnableAlphaTest == 0)
zelda logo looks great but Mystical Ninja text has black artifacts again.
however mario 64 text looks good
![GLideN64_SUPER_MARIO_64_017](https://user-images.githubusercontent.com/7278372/68079633-cc46c080-fda9-11e9-9ee2-2455ab56963c.png)

weird that the Mystical Ninja text is testing for alpha, when the text is being rendered fully opaque. anything else I could use to test for opaqueness in the destination?

**Comment by oddMLan** [comment]
This is most likely not worth the hassle. None of these problems are present in native resolution factors. The N64 has a nearly infinite amount of combinations of setting up these display lists, it would be nearly impossible to account for all cases. Something that would address this would to enforce using native resolution factors by default. Like using the first factor greater than the output window.

However this filter is still potentially good for upscaling the window image in situations when the input resolution is smaller than the output resolution. If #2088 gets implemented, it would be nice to have this filter :)

**Comment by Simp2** [comment]
Hey, I was just wondering if OoT's HUD is actually supposed to look this blurry, even on the rev. c790bc53 build? It looks like some bilinear filter is smudging everything, and I prefer a cleaner pixelated look.
![oot3](https://user-images.githubusercontent.com/6455166/68981149-80940e00-07d0-11ea-88d9-8e4c80c91c78.png)
(Taken on the current public release 4.0 at 2x native resolution)

![GLideN64_THE_LEGEND_OF_ZELDA_000](https://user-images.githubusercontent.com/6455166/68981332-06b05480-07d1-11ea-8407-a4335c9b1a7e.png)
(Taken on the rev. c790bc53 build posted by oddMLan running at 4x native resolution (this version at 2x native resolution looks pretty much the same as 4.0 at 2x))

**Comment by oddMLan** [comment]
This build doesn't modify 3-point filtering which is what you're using.

**Comment by Simp2** [comment]
Pardon me for my ignorance on this subject, but were you implying that the build you posted won't affect this particular game/scenario or were you suggesting that I switch from 3-point filtering to the other option I see in the options, bilinear filtering (which I just tried and saw no difference with)? 

I'm confused, because stuff like Mario 64's menus/HUD look clean on your build, as you demonstrated in the screenshots you posted earlier, but OoT still has that fuzzy HUD. I'm going to assume that these games handle these elements differently.

**Comment by oddMLan** [comment]
>I'm going to assume that these games handle these elements differently.

Yes, that's correct. OOT mandates the HUD to be filtered, hence the blurriness. SM64's HUD ignores the filtering setting, so the pixels stay the same. There's currently no way to disable the filtering because GLideN64 filters the same way as the N64 dictates. There's already a ticket asking for a setting to disable all texture filtering: #574
but this will make landscape textures look pixelated too.

**Comment by Simp2** [comment]
I see now, thanks for your explanation.

**Comment by eqagunn** [comment]
> Test build without commit [0338f27](https://github.com/gonetz/GLideN64/commit/0338f272254f58d0940e5421412642027dffdcc7) as requested by @DonelBueno
> I confirm that this also gets rid of bug #2112
> [GLideN64.zip](https://github.com/gonetz/GLideN64/files/3800935/GLideN64.zip)

Any chance you could provide Mupen64Plus compatible version?
